tests/levenberg_marquardt_test.py

```diff
@@ -0,0 +1,149 @@
+
+from jaxopt import LevenbergMarquardt
+
+import numpy as onp
+
+
+def _rosenbrock(x):
+  return jnp.sqrt(2) * jnp.array([10 * (x[1] - x[0]**2), (1 - x[0])])
+
+
+def _double_exponential(coeffs, x):
+  return coeffs[0] * jnp.exp(-coeffs[2] * x) + coeffs[1] * jnp.exp(
+      -coeffs[3] * x)
+
+
+def _double_exponential_residual(coeffs, x, y):
+  return y - _double_exponential(coeffs, x)
+
+
+def _double_exponential_residual_jac(coeffs, x, y, eps=1e-5):
+  """Return the numerical Jacobian."""
+  lm = LevenbergMarquardt(
+      residual_fun=_double_exponential_residual,
+      damping_parameter=1e-3,
+      xtol=1e-14,
+      gtol=1e-8,
+      maxiter=100,)
+
+  # Sets eps only at idx, the rest is zero
+  eps_at = lambda idx: onp.array([int(i == idx)*eps for i in range(len(x))])
+
+  res1 = jnp.zeros((len(coeffs), len(x)))
+  res2 = jnp.zeros((len(coeffs), len(x)))
+  for i in range(len(x)):
+    res1 = res1.at[:,i].set(lm.run(coeffs, x + eps_at(i), y).params)
+    res2 = res2.at[:,i].set(lm.run(coeffs, x - eps_at(i), y).params)
+
+  twoeps = 2 * eps
+  return (res1 - res2) / twoeps
+
+
+class LevenbergMarquardtTest(jtu.JaxTestCase):
+
+  def setUp(self):
+    super().setUp()
+
+    # Exponential fit data from immoptibox "A MATLAB TOOLBOX FOR OPTIMIZATION
+    # AND DATA FITTING" by Hans Bruun Nielsen
+    self.double_exponential_x = onp.linspace(0.02, 0.9, 45)
+    self.double_exponential_y = onp.array([
+        0.090542, 0.124569, 0.179367, 0.195654, 0.269707, 0.286027, 0.289892,
+        0.317475, 0.308191, 0.336995, 0.348371, 0.321337, 0.299423, 0.338972,
+        0.304763, 0.288903, 0.30082, 0.303974, 0.283987, 0.262078, 0.281593,
+        0.267531, 0.218926, 0.225572, 0.200594, 0.197375, 0.18244, 0.183892,
+        0.152285, 0.174028, 0.150874, 0.12622, 0.126266, 0.106384, 0.118923,
+        0.091868, 0.128926, 0.119273, 0.115997, 0.105831, 0.075261, 0.068387,
+        0.090823, 0.085205, 0.067203
+    ])
+
+  # This unit test is for checking rosenbrock function based on the example
+  # provided by K. Madsen & H. B. Nielsen in the book "Introduction to
+  # Optimization and Data Fitting" Example 6.9. page 123.
+  def test_rosenbrock(self):
+
+    lm = LevenbergMarquardt(
+        residual_fun=_rosenbrock,
+        damping_parameter=1e-3,
+        xtol=1e-12,
+        gtol=1e-8,
+        maxiter=16)
+    optimize_info = lm.run(onp.array([-1.2, 1]))
+
+    self.assertArraysAllClose(
+        optimize_info.params - onp.array([1, 1]),
+        onp.array([-4.07e-09, -8.16e-09]),
+        atol=1e-11)
+    self.assertAllClose(optimize_info.state.iter_num, 16)
+
+  # This unit test is for checking rosenbrock function based on the example
+  # provided by K. Madsen & H. B. Nielsen in the book "Introduction to
+  # Optimization and Data Fitting" Example 6.10-. page 124. Data taken from
+  # immoptibox "A MATLAB TOOLBOX FOR OPTIMIZATION AND DATA FITTING" H.B. Nielsen
+  @parameterized.product(solver=['cg', 'bicgstab', 'gmres', 'cholesky', 'inv'])
```

Comment:
Maybe we should only accept 'cholesky', 'inv' and Callable. This way we don't have to test all possible solvers (we try to keep tests short when possible).

Reply:
Sure :).